1주차: 스터디 (#2)
fe-deep-dive/study-react-deep-dive: week01/01장 리액트 개발을 위해 꼭 알아야 할 자바스크립트 2.md

map과 달리 forEach는 아무런 반환값이 없다. (undefined로 의미 없다.)

forEach는 실행하는 순간 에러를 던지거나 프로세스를 종료하지 않는 이상 멈출 수 없다. break, return, 그 무엇을 이용해도 배열 순회를 멈출 수 없다.

```jsx
function run() {
  const arr = [1, 2, 3];
  arr.forEach((item) => {
    console.log(item);
    if (item === 1) {
      console.log('finished!');
      return; // 함수의 return이 아닌 콜백 함수의 return 문이므로 순회는 계속된다.
    }
  });
}

run();
// 1
// finished!
// 2
// 3
```

forEach 내부의 콜백 함수는 무조건 O(n)만큼 실행되므로 코드 작성과 실행 시에 최적화할 가능성이 있는지 검토하자.

### 1.6.5 삼항 조건 연산자

```jsx
function Component({condition}) {
  return <>{condition ? '참' : '거짓'}</>
}
```

삼항 조건 연산자는 기존의 if 조건문을 간단하게 쓸 수 있어 리액트 내부에서 조건부로 JSX를 렌더링하기 위해 널리 쓰인다.

삼항 연산자는 가독성을 위해 중첩해서 쓰지 않는 편이 좋다.

**삼항 연산자 대신 조건부 렌더링 구현하기**

JSX 내부에서 자바스크립트 할당식을 사용해 구현할 수 있다. 그러나 JSX 내부의 가독성을 해치고 불필요하게 즉시 실행 함수를 선언해야 하므로 선호되지 않는다.

```jsx
import { useState } from 'react';

export default function App() {
  const [color, setColor] = useState('');
  return (
    <div>
      {(() => {
        if (color === 'red') {
          return '빨간색이다.';
        } else {
          return ' 빨간색이 아니다.';
        }
      })()}
    </div>
  );
}
```

### 1.6.6 정리

리액트 애플리케이션을 작성할 때 만나게 되는 주요 문법과 메서드에 대해 알아봤다.

> - 자바스크립트 코드에 관심이 많다면 ECMAScript의 신규 문법을 제안하는 저장소에 방문해보자. https://github.com/c39/proposal

Reviewer: ```suggestion
- 자바스크립트 코드에 관심이 많다면 ECMAScript의 신규 문법을 제안하는 저장소에 방문해보자. https://github.com/tc39/proposals
```

- 최신 문법을 리액트에 반영하기로 마음먹었다면 바벨과 같은 도구를 이용한 트랜스파일을 지원하는지, 사용자의 디바이스에서 별도 조치 없이 사용 가능한지 점검하자.

## 1.7 선택이 아닌 필수, 타입스크립트
Deno, Bun 등 Node.js의 대항마로 출시되는 런타임들도 타입스크립트를 기본으로 지원한다. 이제 현업에서도 대다수의 프로젝트는 타입스크립트로 개발되고 있다.

타입스크립트를 도입하면 빌드 타임에서 타입을 체크하여 더욱 안전하게 작성하고 잠재적인 버그도 크게 줄일 수 있어 타입스크립트는 선택이 아닌 필수이다.

### 1.7.1 타입스크립트란?

>  "TypesScrip is Javascript with syntax for types." 기존 자바스크립트 문법에 타입을 가미한것이 바로타입스크립 트라할수있다.

Reviewer: ```suggestion
 "TypesScript is Javascript with syntax for types." 기존 자바스크립트 문법에 타입을 가미한 것이 바로 타입스크립트라 할 수 있다.
```

Author: 아오 감사합니돠 사랑햄 =^.^=

  // can't be used to index type 'Hello'.
  // No index signature with a parameter fo type 'string' was found on type 'Hello'
  const value = hello[key];
  return value;
});
```

Object.keys(hello)는 string[] 타입을 반환하므로 string은 hello의 인덱스 키로 접근할 수 없다. 이를 해결하기 위한 방법은 다양하다. 

```jsx
// 타입 단언을 통해 해결하기

Object.keys(hello as Array<keyof Hello>).map((key) => {
  const value = hello[key];
  return value;
});

// Object.keys에 대한 반환 타입을 개발자가 단언한 타입으로 강제하기
// 타입가드 함수를 만드는방법
function keysOf<T extends Object>(obj: T): Array<keyof T> {
  return Array.from(Object.keys(obj)) as Array<keyof T>;
}
keysOf(hello).map((key) => {
  const value = hello[key];
  return value;
});

// 가져온 key를 단언하는 방법
Object.keys(hello).map((key) => {
  const value = hello[key as keyof hello];
  return value;
});
```

> 타입스크립트는 왜 Object.keys를 string[]으로 강제했을까?
> 자바스크립트는 덕 타이핑으로 객체를 비교해야 한다.
> 즉, 어떤 객체가 필요한 변수와 메서드만 지니고 있다면 그냥 해당 타입에 속하도록 인정해 주는 것이다.
> 이를 반영해 타입스크립트는 타입 검사를 할 때 값의 형태에 초점을 맞추는 것이다.

### 1.7.3 타입스크립트 전환 가이드

점진적으로 타입스크립트로 전환하는 방법을 알아보자.

**tsconfig.json 작성하기**
```json
{
  "compilerOptions": { "outDir": "./dist", "allows": true, "target": "es5" },
  "include": ["./src/**/*"]
}
```

> - outbir: .ts나 .js가 만들어진 결과를 넣어두는 폴더다. tsc는 타입스크립트를 자바스크립트로 변환하는 명령어인데, 이 tsc를 사용하면 결과물이 outDir로 넘어간다.

Reviewer: ```suggestion
- outDir: .ts나 .js가 만들어진 결과를 넣어두는 폴더다. tsc는 타입스크립트를 자바스크립트로 변환하는 명령어인데, 이 tsc를 사용하면 결과물이 outDir로 넘어간다.
```

- allowJs: .js 파일을 허용할 것인지 여부다. 자바스크립트가 존재하는 과도기적인 프로젝트이므로 true로 설정해둔다.
- target: 결과물이 될 자바스크립트 버전을 지정한다.
- include: 트랜스파일할 자바스크립트와 타입스크립트 파일을 지정한다.

**JSDoc과 @ts-check을 활용해 점진적으로 전환하기**

타입스크립트로 전환하지 않더라도 타입을 체크하는 방법이 있다. 파일 상단에 //@ts-check를 선언하고 JSDoc을 활용해 변수나 함수에 타입을 제공하면 타입스크립트 컴파일러가 자바스크립트 파일의 타입을 확인한다.

손이 가는 작업이므로 JSDoc을 이미 사용했거나 타입스크립트로 전환하기 어려울 때만 하는 것을 추천한다. 그 외에는 바로 .ts로 파일 확장자를 변경하고 바로 작업하는 편이 더 빠르게 전환할 수 있다.

**타입 기반 라이브러리 사용을 위해 @types 모듈 설치하기**

> @types라 불리는 DefinitelyTyped를 설치해 타입스크립트로 작성되지 않은 코드에 대해 타입을 제공할 수 있다. 만약 파일을 .ts로 전환했는데 import에 에 "Cannot find module lodash or its corresponding type declarations'라는 오류메시지가 출력된다면 @types 라이브러리를 설치해야한다. 이 에러는 Lodas라는 라이브러리 내부에서 별도의 d.ts와 같은 타입 파일을 제공하지 않기 때문에 발생한다. @types를 검색해 별도 타입을 제공하는 라이브러리가 있는지 확인하고 설치하자. 

Reviewer: ```suggestion
@types라 불리는 DefinitelyTyped를 설치해 타입스크립트로 작성되지 않은 코드에 대해 타입을 제공할 수 있다. 만약 파일을 .ts로 전환했는데 import에 에 "Cannot find module lodash or its corresponding type declarations'라는 오류메시지가 출력된다면 @types 라이브러리를 설치해야한다. 이 에러는 Lodash라는 라이브러리 내부에서 별도의 d.ts와 같은 타입 파일을 제공하지 않기 때문에 발생한다. @types를 검색해 별도 타입을 제공하는 라이브러리가 있는지 확인하고 설치하자. 
```

> Next.js와 같이 비교적 최근에 만들어진 라이브러리는 이미 자체적으로 타입스크립트 지원 기능이 라이브러리에 내장되어 있다.

> **파일 단위로 조금씩 전환하기**

Reviewer: ```suggestion
**파일 단위로 조금씩 전환하기**

```
